### PR TITLE
Add copy-to-clipboard buttons to offer templates

### PR DESCRIPTION
**Pull request by mnutt**
sandstorm-clipboard

Adds a {clipboardButton: 'left'} or {clipboardButton: 'right'} options to display the copy-to-clipboard button on one side or the other. It uses clipboard.js (inlined in offertemplate.html, let me know if you'd prefer a better way) which uses HTML5 clipboard api and falls back to highlighting the text for the user to hit "copy".

### CODE REVIEW

**Comment by garply** [comment]
Hey admins! Can one of you let me know if it's safe to build this pull request? See https://alpha.sandstorm.io/grain/kB7bofATTL2jGuKSFwmqdA/GithubPRCommands for a list of commands.

**Comment by paulproteus** [comment]
Garply, ok to test

**Comment by paulproteus** [comment]
@zarvox probably wants to review this.

As a side note, https://github.com/sandstorm-io/sandstorm/wiki/Giving-credit-to-other-open-source-projects describes my current understanding of how we attribute other open source projects, so if this gets merged, we should follow those steps (or change the process).

**Comment by paulproteus** [comment]
I am personally excited about seeing this clipboard functionality land in Sandstorm.

Additionally, we should presumably document it somewhere like https://docs.sandstorm.io/en/latest/developing/http-apis/#creating-an-offer-template - I'll let @zarvox provide feedback on the feature before asking for docs, but I wanted to bring that up at least.

**Comment by kentonv** [comment]
This looks awesome. Agreed @zarvox should review (when he gets back).

File: shell/public/clippy.svg [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<svg height="16" width="14" xmlns="http://www.w3.org/2000/svg">
+  <path d="M2 12h4v1H2v-1z m5-6H2v1h5v-1z m2 3V7L6 10l3 3V11h5V9H9z m-4.5-1H2v1h2.5v-1zM2 11h2.5v-1H2v1z m9 1h1v2c-0.02 0.28-0.11 0.52-0.3 0.7s-0.42 0.28-0.7 0.3H1c-0.55 0-1-0.45-1-1V3c0-0.55 0.45-1 1-1h3C4 0.89 4.89 0 6 0s2 0.89 2 2h3c0.55 0 1 0.45 1 1v5h-1V5H1v9h10V12zM2 4h8c0-0.55-0.45-1-1-1h-1c-0.55 0-1-0.45-1-1s-0.45-1-1-1-1 0.45-1 1-0.45 1-1 1h-1c-0.55 0-1 0.45-1 1z" />
```
Comment (zarvox) [inline shell/public/clippy.svg]: What's the source of this `clippy.svg`?  I'm not seeing it in https://github.com/zenorocha/clipboard.js

File: shell/public/offer-template.html [hunk 2]
```diff
@@ -10,20 +10,89 @@
...
...lveOptions(r),this.listenClick(n)}return i(e,t),e.prototype.resolveOptions=function t(){var e=arguments.length<=0||void 0===arguments[0]?{}:arguments[0];this.action="function"==typeof e.action?e.action:this.defaultAction,this.target="function"==typeof e.target?e.target:this.defaultTarget,this.text="function"==typeof e.text?e.text:this.defaultText},e.prototype.listenClick=function t(e){var n=this;this.listener=d.default(e,"click",function(t){return n.onClick(t)})},e.prototype.onClick=function t(e){var n=e.delegateTarget||e.currentTarget;this.clipboardAction&&(this.clipboardAction=null),this.clipboardAction=new s.default({action:this.action(n),target:this.target(n),text:this.text(n),trigger:n,emitter:this})},e.prototype.defaultAction=function t(e){return a("action",e)},e.prototype.defaultTarget=function t(e){var n=a("target",e);return n?document.querySelector(n):void 0},e.prototype.defaultText=function t(e){return a("text",e)},e.prototype.destroy=function t(){this.listener.destroy(),this.clipboardAction&&(this.clipboardAction.destroy(),this.clipboardAction=null)},e}(l.default);n.default=h,e.exports=n.default},{"./clipboard-action":8,"good-listener":5,"tiny-emitter":7}]},{},[9])(9)});
```
Comment (zarvox) [inline shell/public/offer-template.html]: I kinda wish there were a better way to do this than inlining the minified `clipboard.js`.  I guess there's not a great solution short of adding some sort of build step to include, which seems like a big ask.

For now, could you add a comment with a link to the source here, so it's clearer what this is and where it came from?

We'll also need to add appropriate attribution for `clipboard.js` and `clippy.svg` in `LICENSE` - could you do that following the general pattern in that file?

File: shell/public/offer-template.html [hunk 3]
```diff
@@ -10,20 +10,89 @@
...
+  right: 20px;
+}
     </style>
   </head>
   <body>
-    <pre id="text"></pre>
+    <div id="container">
+      <pre id="text"></pre>
+    </div>
+    <div id="clipboard" data-clipboard-target="#text"></div>
```
Comment (zarvox) [inline shell/public/offer-template.html]: I'm not sure how accessible this `<div>` is - can it be made into a `<button>` with an accessible label like "Copy" so it'll appear in the `taborder` and so folks like Nolan and other keyboard users will be able to take advantage of this?

**Comment by zarvox** [comment]
Big :+1: on the idea.  Thanks for building this!

I'd appreciate a couple additions for license reasons, and if you can make the button more accessible, that would be fantastic. :)

**Comment by mnutt** [comment]
Ok, I've updated the LICENSE and switched to a `<button>`, although I realized I'm actually not quite sure how to get it into the `taborder`, especially since it's in an iframe. @zarvox do you know what attribute I'd add?

Also, I moved clipboard.js to an external file. If we started accumulating more js files we'd probably need some sort of build/concat step, but as it is I think it's ok?

File: shell/public/offer-template.html [hunk 4]
```diff
@@ -10,20 +10,92 @@
...
+  right: 20px;
+}
     </style>
   </head>
   <body>
-    <pre id="text"></pre>
+    <div id="container">
+      <pre id="text"></pre>
+    </div>
+    <button id="clipboard" label="Copy" data-clipboard-target="#text"></button>
```
Comment (zarvox) [inline shell/public/offer-template.html]: I could be mistaken, but I think you want `aria-label` here, not `label`.  I couldn't find an attribute `label` anywhere in the HTML5 spec.

**Comment by zarvox** [comment]
The iframe itself is in the taborder, and once you tab into the iframe, you'll traverse all of the focusable elements within.  Since `<button>` is added to the taborder by default, I think this works fine.

I'm okay with making a blocking request for script on the first fetch of `offer-template.html`, especially since if the fetch fails it gracefully degrades to the old template's behavior.

So I think this just needs s/label/aria-label/ over by my last inline comment and we're good to go.

**Comment by mnutt** [comment]
Alright, updated. I'm happy to take a stab at the documentation this weekend.

**Comment by kentonv** [comment]
Garply, retest this please

**Comment by mnutt** [comment]
@paulproteus I wrote a bit of documentation, let me know if you think there is anything else to add.

**Comment by paulproteus** [comment]
The only thing I'd add to these docs is an `img src` that shows the clipboard button. That is not a show-stopper.

Honestly, the animated GIF you created in this pull request would be great as part of the docs, too.

Neither is a show-stopper. If you want, you can consider adding the animated GIF to the bottom or the top of this section: "Creating an offer template"

In fact, since I am currently putting images for the docs in a separate Davros grain, and using static publishing to get them published, here, have a URL you can use:

https://alpha-evgl4wnivwih0k6mzxt3.sandstorm.io/offer-template-video-by-mnutt.gif

and you can add it to the docs if you want, and if not, that's OK too.

If you do add it to the docs in here, I will take that as you licensing under the same terms as your other contributions to Sandstorm.

**Comment by paulproteus** [comment]
+1 from me for docs/ changes

**Comment by paulproteus** [comment]
@zarvox I guess we await a refreshed +1 from you.

**Comment by zarvox** [comment]
:+1: from me.  Thanks again!

**Comment by paulproteus** [comment]
Great - @kentonv I believe this is ready for your final review/merge.

**Comment by kentonv** [comment]
Ugh apparently somehow this dropped off my list after I told Garply to retest.

So sorry for another delay, but @zarvox ran an auto-formatter across the codebase and now all our PRs have merge conflicts.

BTW I wonder how @neynah feels about the clipboard icon or if she wants to make a replacement.

Also which browsers has this been tested in? As I understand it different browsers have different rules about when you're allowed to invoke a "copy" command from script.

**Comment by neynah** [comment]
Sorry for the delay, I missed this! I'd like to simplify/redesign the icon & I also think it should consistently be justified right.

I'm OK with using this icon for now until I design a new one this week.

**Comment by zarvox** [comment]
@kentonv per https://clipboardjs.com/ this works in Chrome 42+, Firefox 41+, IE 9+, Opera 29+, and degrades to using the Selection API to select the contents to be copied otherwise.  I think that's perfectly fine, and is an improvement over the status quo, which has neither selection nor clipboard support.

FWIW, @mnutt I'm happy to rebase this atop master, since as far as I'm concerned it's ready to merge and I'm responsible for breaking everything - if you haven't replied or rebased this branch yourself by Friday, I'll just do it as a blur day task.

Thanks again!

**Comment by kentonv** [comment]
@zarvox Be sure to use Nena's clipboard icon.

**Comment by garply** [comment]
Hey admins! Can one of you let me know if it's safe to build this pull request? See https://alpha.sandstorm.io/grain/kB7bofATTL2jGuKSFwmqdA/GithubPRCommands for a list of commands.

**Pull request closed without merging**